File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/HealthCheck.java
package uk.gov.justice.digital.hmpps.keyworker.services;

import lombok.AllArgsConstructor;
import org.springframework.boot.actuate.health.Health;
import org.springframework.boot.actuate.health.HealthIndicator;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;

import java.time.Duration;

import static lombok.AccessLevel.PROTECTED;

@AllArgsConstructor(access = PROTECTED)
public abstract class HealthCheck implements HealthIndicator {
    private final WebClient webClient;
    private final Duration timeout;

    @Override
    public Health health() {
        try {
            final var responseEntity =
                    webClient.get()
                            .uri("/health/ping")
                            .retrieve()
                            .toEntity(String.class)
                            .block(timeout);
            return Health.up().withDetail("HttpStatus", responseEntity.getStatusCode()).build();
        } catch (final WebClientResponseException e) {
            return Health.down(e).withDetail("body", e.getResponseBodyAsString()).build();
        } catch (final Exception e) {
            return Health.down(e).build();
        }
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/keyworker/services/health/HealthCheck.kt
package uk.gov.justice.digital.hmpps.keyworker.services.health

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.health.contributor.Health
import org.springframework.boot.health.contributor.ReactiveHealthIndicator
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono

@Component
class PrisonApiHealth(
  @Qualifier("prisonApiHealthWebClient") webClient: WebClient,
) : WebClientPingHealthIndicator(webClient)

@Component
class ComplexityOfNeedApiHealth(
  @Qualifier("complexityOfNeedHealthWebClient") webClient: WebClient,
) : WebClientPingHealthIndicator(webClient)

open class WebClientPingHealthIndicator(
  private val webClient: WebClient,
) : ReactiveHealthIndicator {
  override fun health(): Mono<Health> =
    webClient
      .get()
      .uri("/health/ping")
      .retrieve()
      .toBodilessEntity()
      .map { response ->
        Health
          .up()
          .withDetail("HttpStatus", response.statusCode.toString().substringAfterLast(" "))
          .build()
      }.onErrorResume { Mono.just(Health.down(it).build()) }
}
